core-rust/state-manager/src/receipt.rs

```diff
@@ -85,63 +94,68 @@ impl From<TransactionOutcome> for LedgerTransactionOutcome {
     pub next_epoch: Option<(BTreeMap<ComponentAddress, Validator>, u64)>,
 }
 
 impl LedgerTransactionReceipt {
-    pub fn get_hash(&self) -> LedgerReceiptHash {
-        LedgerReceiptHash::for_receipt(self)
+    pub fn get_consensus_receipt(&self) -> ConsensusReceipt {
+        ConsensusReceipt {
+            outcome: self.outcome.clone(),
+            substate_change_root: Self::compute_substate_change_root(&self.substate_changes),
+        }
+    }
+
+    fn compute_substate_change_root(substate_changes: &SubstateChanges) -> SubstateChangeHash {
+        // TODO: implement using a merkle tree
+        SubstateChangeHash::from(blake2b_256_hash(scrypto_encode(substate_changes).unwrap()))
     }
 }
 
-impl From<(CommitResult, TransactionExecutionTrace)> for LedgerTransactionReceipt {
+impl From<(CommitResult, TransactionExecutionTrace)> for LocalTransactionReceipt {
     fn from((commit_result, execution_trace): (CommitResult, TransactionExecutionTrace)) -> Self {
         let next_epoch = commit_result.next_epoch();
         Self {
-            outcome: commit_result.outcome.into(),
-            fee_summary: commit_result.fee_summary,
-            fee_payments: commit_result.fee_payments,
-            application_logs: commit_result.application_logs,
-            substate_changes: fix_state_updates(commit_result.state_updates),
-            state_update_summary: commit_result.state_update_summary,
-            resource_changes: execution_trace.resource_changes,
-            next_epoch,
-        }
-    }
-}
-
-impl TryFrom<EngineTransactionReceipt> for LedgerTransactionReceipt {
-    type Error = String;
-
-    fn try_from(engine_receipt: EngineTransactionReceipt) -> Result<Self, Self::Error> {
-        match engine_receipt.result {
-            TransactionResult::Commit(commit_result) => Ok(LedgerTransactionReceipt::from((
-                commit_result,
-                engine_receipt.execution_trace,
-            ))),
-            TransactionResult::Reject(error) => Err(format!(
-                "Can't create a ledger receipt for rejected txn: {error:?}"
-            )),
-            TransactionResult::Abort(result) => Err(format!(
-                "Can't create a ledger receipt for aborted txn: {result:?}"
-            )),
+            on_ledger: LedgerTransactionReceipt {
+                outcome: LedgerTransactionOutcome::from(&commit_result.outcome),
+                substate_changes: fix_state_updates(commit_result.state_updates),
+            },
+            local_execution: LocalTransactionExecution {
+                outcome: commit_result.outcome.into(),
+                fee_summary: commit_result.fee_summary,
+                fee_payments: commit_result.fee_payments,
+                application_logs: commit_result.application_logs,
+                state_update_summary: commit_result.state_update_summary,
+                resource_changes: execution_trace.resource_changes,
+                next_epoch,
```

Comment:
As a heads up - some of these things (eg next_epoch), and eventually balance_changes, state_update_summary, etc - will be computable from the LedgerTransactionReceipt, and so won't need to be explicitly stored 👍

Reply:
> will be computable

wooooo, finally 🎉